File: Data Structures & Algorithms/anagram-groups/submission-3.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String,List<String>> map = new HashMap<>();
        for(int i=0;i<strs.length;i++){
            int[] arr = new int[26];
            for(int j=0;j<strs[i].length();j++){
                arr[strs[i].charAt(j)-'a']++;
            }
            StringBuffer str = new StringBuffer();
            for(int k=0;k<26;k++){
                str.append(k).append("_").append(arr[k]);
            }
            String newS = str.toString();
            if(map.containsKey(newS)){
                List<String> listS = new ArrayList(map.get(newS));
                listS.add(strs[i]);
                map.put(newS,listS);
            }
            else{
                map.put(newS,List.of(strs[i]));
            }
        }
        List<List<String>> ans = new ArrayList<>();
        for(Map.Entry<String,List<String>> entry: map.entrySet()){
            ans.add(entry.getValue());
        }
        return ans;
    }
}

File: Data Structures & Algorithms/string-encode-and-decode/submission-3.cpp
class Solution {
public:

    string encode(vector<string>& strs) {
        string ans;
        for(int i=0;i<strs.size();i++){
            ans += to_string(strs[i].size()) + "#" + strs[i];
        }
        cout<<ans;
        return ans;
    }

    vector<string> decode(string s) {
        int i=0,j=0;
        int n = s.size();
        vector<string> ans;
        while(j<n){
            if(s[j]=='#'){
                // cout<< s.substr(i,j-i) << "\n";
                int size = stoi(s.substr(i,j-i));
                ans.push_back(s.substr(j+1,size));
                j += size ;
                i = j+1;
            }
            j++;
        }
        return ans;
    }
};
